pkg/util/pki/csr.go

```diff
@@ -278,148 +302,213 @@ func buildKeyUsagesExtensionsForCertificate(crt *v1.Certificate) ([]pkix.Extensi
 	} else {
-		cert.Subject = pkix.Name{
-			Country:            subject.Countries,
-			Organization:       organization,
-			OrganizationalUnit: subject.OrganizationalUnits,
-			Locality:           subject.Localities,
-			Province:           subject.Provinces,
-			StreetAddress:      subject.StreetAddresses,
-			PostalCode:         subject.PostalCodes,
-			SerialNumber:       subject.SerialNumber,
-			CommonName:         commonName,
+		cert.MaxPathLen = 0
+		cert.MaxPathLenZero = false
+	}
+
+	// We expect that if the CSR contains a KeyUsage/ ExtendedKeyUsage
+	// extension, these KeyUsages match the usages that we set here. To
+	// be safe, we append the usages instead of replacing them.
+	// This should have been enforced by our validation webhook.
+
+	cert.KeyUsage |= keyUsage
+OuterLoop:
+	for _, usage := range extKeyUsage {
+		for _, existingUsage := range cert.ExtKeyUsage {
+			if existingUsage == usage {
+				continue OuterLoop
+			}
 		}
+
+		cert.ExtKeyUsage = append(cert.ExtKeyUsage, usage)
 	}
 
-	return cert, nil
+	return cert
 }
 
-// GenerateTemplate will create a x509.Certificate for the given
-// CertificateRequest resource
-func GenerateTemplateFromCertificateRequest(cr *v1.CertificateRequest) (*x509.Certificate, error) {
-	certDuration := apiutil.DefaultCertDuration(cr.Spec.Duration)
-	keyUsage, extKeyUsage, err := KeyUsagesForCertificateOrCertificateRequest(cr.Spec.Usages, cr.Spec.IsCA)
+// GenerateTemplateFromCSRPEM will create a x509.Certificate for the
+// given csrPEM.
+// Call OverrideTemplateFromOptions to override the duration, isCA, maxPathLen, keyUsage, and extKeyUsage.
+func GenerateTemplateFromCSRPEM(csrPEM []byte) (*x509.Certificate, error) {
+	csr, err := DecodeX509CertificateRequestBytes(csrPEM)
 	if err != nil {
 		return nil, err
 	}
-	return GenerateTemplateFromCSRPEMWithUsages(cr.Spec.Request, certDuration, cr.Spec.IsCA, keyUsage, extKeyUsage)
-}
-
-func GenerateTemplateFromCSRPEM(csrPEM []byte, duration time.Duration, isCA bool) (*x509.Certificate, error) {
-	var (
-		ku  x509.KeyUsage
-		eku []x509.ExtKeyUsage
-	)
-	return GenerateTemplateFromCSRPEMWithUsages(csrPEM, duration, isCA, ku, eku)
-}
 
-func GenerateTemplateFromCSRPEMWithUsages(csrPEM []byte, duration time.Duration, isCA bool, keyUsage x509.KeyUsage, extKeyUsage []x509.ExtKeyUsage) (*x509.Certificate, error) {
-	block, _ := pem.Decode(csrPEM)
-	if block == nil {
-		return nil, errors.New("failed to decode csr")
+	if err := csr.CheckSignature(); err != nil {
+		return nil, err
 	}
 
-	csr, err := x509.ParseCertificateRequest(block.Bytes)
+	return GenerateTemplateFromCSR(csr)
+}
+
+// GenerateTemplateFromCSRPEMWithOptions decodes the given CSR and applies the
+// provided options on top of the decoded template.
+func GenerateTemplateFromCSRPEMWithOptions(
+	csrPEM []byte,
+	duration time.Duration,
+	isCA bool,
+	maxPathLen *int,
+	keyUsage x509.KeyUsage,
+	extKeyUsage []x509.ExtKeyUsage,
+) (*x509.Certificate, error) {
```

Comment:
Instead of an `GenerateTemplateFromCSROptions struct`, I opted for a set of mutator functions. These replace the `OverrideTemplateFromOptions` function.